wakuv2/waku.go

```diff
@@ -632,41 +639,66 @@ func (w *Waku) runFilterMsgLoop() {
 		return
 	}
 
+	// Use it to ping filter peer(s) periodically
+	ticker := time.NewTicker(time.Duration(w.cfg.KeepAliveInterval) * time.Second)
+	defer ticker.Stop()
+
 	for {
 		select {
 		case <-w.quit:
 			return
-		case env, ok := <-w.filterMsgChannel:
-			if ok {
-				envelopeErrors, err := w.OnNewEnvelopes(env, common.RelayedMessageType)
-				// TODO: should these be handled?
-				_ = envelopeErrors
-				_ = err
+		case <-ticker.C:
+			for f, subMap := range w.filterSubscriptions {
+				for id, sub := range subMap {
+					err := w.node.FilterLightnode().IsSubscriptionAlive(context.Background(), sub)
+					if err != nil {
+						w.filterPeerDisconnectMap[sub.PeerID] = time.Now().Unix()
+						delete(subMap, id)
+						// TODO Is it necessary to unsubscribe?
+
+						// Re-subscribe
+						peers := w.findFilterPeers()
+						if len(peers) > 0 {
+							contentFilter := w.buildContentFilter(f.Topics)
+							subDetails, err := w.node.FilterLightnode().Subscribe(context.Background(), contentFilter, filter.WithPeer(peers[0]))
+							if err != nil {
+								w.logger.Warn("could not add wakuv2 filter for peer", zap.Any("peer", peers[0]))
+							}
+
+							w.filterSubscriptions[f][subDetails.ID] = subDetails
+
+						}
+					}
+				}
+			}
+		default:
+			for _, subMap := range w.filterSubscriptions {
+				for _, sub := range subMap {
+					select {
+					case env, ok := <-sub.C:
+						if ok {
+							envelopeErrors, err := w.OnNewEnvelopes(env, common.RelayedMessageType)
+							// TODO: should these be handled?
+							_ = envelopeErrors
+							_ = err
+						}
+					default:
+						// Use an empty default so that reading from each sub.C is non-blocking
+					}
+				}
 			}
 		}
 	}
```

Comment:
I think it would be more performant to remove this `default` and do the filter subscription reading by creating a separate goroutine for each filter subscribtion created (in line 668). Something +- like this:

```go
go func(sub *FilterSubscription){
    for {
          select {
              case <-w.quit:
                   return
              case env, ok := <-sub.C:
                   if ok {
                                  envelopeErrors, err := w.OnNewEnvelopes(env, common.RelayedMessageType)
                                  // TODO: should these be handled?
                                  _  = envelopeErrors
                                  _  = err
                            }
          }
    }

}(subDetails)
```
because then you wouldnt have to worry about reading the channel being blockng and you'd have independent 'workers' reading envelopes from the subscriptions.

Reply:
Yes, thanks, agree, done.

Reply:
Added a test for filter re-subscriptions.